infrastructure/src/test/java/org/corfudb/infrastructure/CorfuLayoutServerTest.java

```diff
@@ -0,0 +1,710 @@
+        // The PROPOSE_LAYOUT_REQUEST should be rejected due to wrong propose rank
+        assertFalse(proposeResponse.getPayload().getProposeLayoutResponse().getProposed());
+        // The LayoutServer should send back the expected phase1rank
+        assertEquals(proposeResponse.getPayload().getProposeLayoutResponse().getRank(), phase1Rank);
+    }
+
+    /**
+     * Test that the LayoutServer correctly rejects a PROPOSE_LAYOUT_REQUEST when
+     * the rank in PROPOSE_LAYOUT_REQUEST is equal to the current phase 2 rank.
+     */
+    @Test
+    public void testProposeDuplicateMessage() throws IOException {
+        Layout l = getDefaultLayout();
+        long payloadEpoch = 0L;
+        long phase1Rank = 5L;
+
+        RequestMsg proposeRequest = getRequestMsg(
+                getBasicHeader(false, true),
+                getProposeLayoutRequestMsg(payloadEpoch, phase1Rank, l)
+        );
+
+        ArgumentCaptor<ResponseMsg> responseCaptor = ArgumentCaptor.forClass(ResponseMsg.class);
+
+        when(mockServerContext.getCurrentLayout()).thenReturn(l);
+        when(mockDataStore.get(argThat(new Phase1Matcher(payloadEpoch))))
+                .thenReturn(new Rank(phase1Rank, DEFAULT_UUID));
+        when(mockDataStore.get(argThat(new Phase2Matcher(payloadEpoch))))
+                .thenReturn(new Phase2Data(new Rank(phase1Rank, DEFAULT_UUID), l));
+        layoutServer.handleMessage(proposeRequest, mockChannelHandlerContext, mockServerRouter);
+        verify(mockDataStore, never()).put(any(), any());
+        verify(mockServerRouter).sendResponse(responseCaptor.capture(), any(ChannelHandlerContext.class));
+
+        ResponseMsg proposeResponse = responseCaptor.getValue();
+        // Assert that the payload has a LAYOUT_RESPONSE and that the base
+        // header fields have remained the same
+        assertTrue(compareBaseHeaderFields(proposeRequest.getHeader(), proposeResponse.getHeader()));
+        assertTrue(proposeResponse.getPayload().hasProposeLayoutResponse());
+        // The PROPOSE_LAYOUT_REQUEST should be rejected because the rank
+        // in PROPOSE_LAYOUT_REQUEST is equal to the current phase 2 rank.
+        assertFalse(proposeResponse.getPayload().getProposeLayoutResponse().getProposed());
+    }
+
+    /**
+     * Test that the LayoutServer correctly acknowledges a COMMIT_LAYOUT_REQUEST.
+     */
+    @Test
+    public void testCommitAck() throws IOException {
+        Layout l = getDefaultLayout();
+        long payloadEpoch = 0L;
+        RequestMsg request = getRequestMsg(
+                getBasicHeader(false, true),
+                getCommitLayoutRequestMsg(false, payloadEpoch, l)
+        );
+
+        ArgumentCaptor<ResponseMsg> responseCaptor = ArgumentCaptor.forClass(ResponseMsg.class);
+        ArgumentCaptor<Layout> layoutCaptor = ArgumentCaptor.forClass(Layout.class);
+        ArgumentCaptor<Long> serverEpochCaptor = ArgumentCaptor.forClass(Long.class);
```

Comment:
Are these argument captors necessary? I think calling verify with the expected object should suffice here. In most cases, if you know the result ahead of time, you don't need an argument captor.